File: php/life.php
<?php

error_reporting(E_ALL);

/*
 *  Our main file should:
 *      Read arguments to get our starting point configuration
 *      Read INI file to initialise rules, world size and layout
 *      Initialise world per settings
 *      Populate world randomly or per specified starting layout
 *      Commence looping:
 *          Calculate new world position (with timing)
 *          Display new world position (with timing)
 *          Display timings (current and total)
 */

/*
 *  Required:
 *      World class to handle the World initialisation, calculation, display
 *          Setup class to handle all initial setup
 *          Cell class to handle individual cell operations (??)
 *      Timer class to handle timing calculations etc
 *      MAIN loop (does this need to be a class?)
 */

require_once 'life/world.php';
require_once 'life/timer.php';

$c_timer = new Timer;
$d_timer = new Timer;

function it_lives()
{
    global $c_timer, $d_timer;

    $world = new World;
    echo $world->to_string();
    display_timings($world->step());

     while(true) {
        $c_timer->start();
        $world->calculate();
        $c_timer->stop();

        $d_timer->start();
        echo $world->to_string();
        $d_timer->stop();

        display_timings($world->step());
    }
}

function display_timings($step)
{
    global $c_timer, $d_timer;

    echo "#$step | Calc " . $c_timer->to_string()." | Disp ".$d_timer->to_string()."\n";
}

it_lives();

File: go/life.go
package main

import (
	"fmt"
	"log"
	"os"

	"./life"
)

func main() {
	file, err := os.OpenFile("../life-go.log", os.O_CREATE|os.O_APPEND, 0644)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()
	log.SetOutput(file)
	log.Print("Game of Life: Go version")
	var cTimer, dTimer life.Timer
	var world life.World

	world.Init("default", "conway", true)
	fmt.Print(world.Render(false))
	displayTimings(cTimer, dTimer, world.Step(), false)

	for world.Step() < 1000 || !cTimer.IsAverageStable(20) {
		cTimer.Start()
		world.Calculate()
		cTimer.Stop()

		dTimer.Start()
		fmt.Print(world.Render(false))
		dTimer.Stop()

		displayTimings(cTimer, dTimer, world.Step(), false)
	}
	displayTimings(cTimer, dTimer, world.Step(), true)
}

func displayTimings(cTimer, dTimer life.Timer, step int64, toLog bool) {
	str := fmt.Sprintf("Calc %s | Disp %s", cTimer.String(), dTimer.String())
	fmt.Printf("#%d | %s\n", step, str)
	if toLog {
		log.Printf("Stable at %d passes: %s", step, str)
	}
}
